scrapy/addons/__init__.py

```diff
@@ -0,0 +1,270 @@
+    @classmethod
+    def search_addon(cls, path):
+        """Search for an add-on by its path and return it.
+
+        If no object is found at ``path``, it is tried again first with
+        ``projectname.addons`` prepended (pointing to the current project's
+        ``addons`` folder), then with ``scrapy.addons`` prepended (poiting to
+        Scrapy's built-in add-ons).
+
+        :param path: Python path to an add-on
+        :type path: ``str``
+        """
+        prefixes = ['', 'scrapy.addons.']
+        try:
+            prefixes.insert(1, get_project_path() + '.addons.')
+        except NotConfigured:
+            warnings.warn("Unable to locate project Python path")
+        for prefix in prefixes:
+            fullpath = prefix + path
+            try:
+                return load_module_or_object(fullpath)
+            except NameError:
+                pass
+        else:
+            raise NameError("Could not find add-on '%s'" % path)
+
+    def load_cfg(self, cfg=None):
+        """Load add-ons and configurations from given ``ConfigParser`` object or
+        config file path.
+
+        Each add-on should have its own section, where the section has a name in
+        the form ``addon:my_addon_path``. The add-on object is searched for via
+        the :meth:`get_addon` method, i.e. either there is a ``_source``
+        configuration setting or ``my_addon_path`` is a Python path.
+
+        If ``cfg`` is ``None``, ``scrapy.cfg`` will be used.
+
+        :param cfg: ``ConfigParser`` object or config file path from which to \
+            read add-on configuration
+        :type cfg: ``ConfigParser`` or ``str``
+        """
+        if cfg is None:
+            cfg = get_config()
+        elif isinstance(cfg, six.string_types):
+            cfg = config_from_filepath(cfg)
+        for secname in cfg.sections():
+            if secname.startswith("addon:"):
+                addonkey = secname.split("addon:", 1)[1]
+                addoncfg = dict(cfg.items(secname))
+                self.add(addonkey, addoncfg, addonkey)
+
+    def load_module(self, mod=None):
+        """Load add-ons and configurations from given Python module.
+
+        Each add-on must be configured in a Python dictionary with a name in the
+        form ``addon_my_addon_name``. The add-on object is searched for via the 
+        :meth:`get_addon` method, i.e. either there is a ``_source``
+        configuration setting or ``my_addon_name`` is a Python path. As there
+        can be no dots in variable names, you can also specify a ``_name``
+        configuration setting that will subsequently be used instead of
+        ``my_addon_name``.
+
+        If ``mod`` is ``None``, the Scrapy project settings module will be used.
+
+        :param mod: Module from which to read add-on configuration
+        :type mod: Python module (object) or Python path to a module
+        """
+        if mod is None:
+            mod = os.environ.get('SCRAPY_SETTINGS_MODULE')
```

Comment:
Does this mean we can't set addons configuration via other settings sources like command line or custom_settings right? Not sure about the implementation, but couldn't we add a `load_settings()` method which takes an already populated Settings instance and loads the addons config from there? Though I guess we have to discuss the config locally first.